sentry/src/analytics.rs

```
@@ -0,0 +1,82 @@
+
+    for (event, event_payout) in events_with_payouts {
+        let (publisher, ad_unit, referrer, ad_slot, ad_slot_type) = {
+            let (publisher, event_ad_unit, referrer, ad_slot) = match event {
+                Event::Impression {
+                    publisher,
+                    ad_unit,
+                    referrer,
+                    ad_slot,
+                } => (publisher, ad_unit, referrer, ad_slot),
+                Event::Click {
+                    publisher,
+                    ad_unit,
+                    referrer,
+                    ad_slot,
+                } => (publisher, ad_unit, referrer, ad_slot),
+            };
+
+            let ad_unit = event_ad_unit.and_then(|ipfs| {
+                campaign
+                    .clone() // TODO: Remove clone
+                    .ad_units
+                    .into_iter()
+                    .find(|ad_unit| ad_unit.ipfs == ipfs)
+            });
+            let ad_slot_type = ad_unit.as_ref().map(|unit| unit.ad_type.clone());
+            (publisher, ad_unit, referrer, ad_slot, ad_slot_type)
+        };
+
+        let hostname = match (&referrer, session.referrer_header.as_ref()) {
+            (Some(referrer), None) | (None, Some(referrer)) | (Some(referrer), Some(_)) => referrer
+                .split('/')
+                .collect::<Vec<&str>>()
+                .get(2)
+                .map(|hostname| hostname.to_string()),
+
+            (None, None) => None,
+        };
+
+        // DB: Insert or Update all events
+        let event_for_db = EventAnalytics {
+            time,
+            campaign_id: campaign.id,
+            ad_unit,
+            ad_slot,
+            ad_slot_type,
+            advertiser: campaign.creator,
+            publisher,
+            hostname,
+            country: country.to_owned(),
+            os_name: os_name.to_owned(),
+        };
+        let payout_amount = event_payout.values().sum::<Option<UnifiedNum>>().unwrap(); // TODO: Remove
```

- [x] Please move this somewhere at the beginning of the function. This will make sure that we fail fast rather than doing unnecessary work and then failing.
- [x] Include the `payout` in the `EventAnalytics` instead of sending it separately